hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/common/table/read/TestDeleteRecordLogic.scala

```diff
@@ -0,0 +1,182 @@
+
+    val updateData = Seq(
+      (11, "1", "rider-X", "driver-X", 19.10, 9),
+      (9, "2", "rider-Y", "driver-Y", 27.70, 7))
+    val updates = spark.createDataFrame(updateData).toDF(columns: _*)
+    updates.write.format("hudi").
+      option(RECORDKEY_FIELD.key(), "key").
+      option(PRECOMBINE_FIELD.key(), "ts").
+      option(TABLE_TYPE.key(), tableType).
+      option(OPERATION.key(), "upsert").
+      options(opts).
+      mode(SaveMode.Append).
+      save(basePath)
+
+    val deletesData = Seq((-5, "4", "rider-D", "driver-D", 34.15, 6))
+    val deletes = spark.createDataFrame(deletesData).toDF(columns: _*)
+    deletes.write.format("hudi").
+      option(RECORDKEY_FIELD.key(), "key").
+      option(PRECOMBINE_FIELD.key(), "ts").
+      option(TABLE_TYPE.key(), tableType).
+      option(OPERATION.key(), "delete").
+      options(opts).
+      mode(SaveMode.Append).
+      save(basePath)
+
+    val secondUpdateData = Seq(
+      (14, "5", "rider-Z", "driver-Z", 17.85, 3),
+      (-10, "4", "rider-DD", "driver-DD", 34.15, 5))
+    val secondUpdates = spark.createDataFrame(secondUpdateData).toDF(columns: _*)
+    secondUpdates.write.format("hudi").
+      option(RECORDKEY_FIELD.key(), "key").
+      option(PRECOMBINE_FIELD.key(), "ts").
+      option(TABLE_TYPE.key(), tableType).
+      option(OPERATION.key(), "upsert").
+      options(opts).
+      mode(SaveMode.Append).
+      save(basePath)
+
+    val secondDeletesData = Seq(
+      (10, "4", "rider-D", "driver-D", 34.15, 6),
+      (0, "1", "rider-X", "driver-X", 19.10, 8))
+    val secondDeletes = spark.createDataFrame(secondDeletesData).toDF(columns: _*)
+    secondDeletes.write.format("hudi").
+      option(RECORDKEY_FIELD.key(), "key").
+      option(PRECOMBINE_FIELD.key(), "ts").
+      option(TABLE_TYPE.key(), tableType).
+      option(OPERATION.key(), "delete").
+      options(opts).
+      mode(SaveMode.Append).
+      save(basePath)
+
+    val thirdUpdateData = Seq((-8, "4", "rider-DDD", "driver-DDD", 20.00, 1))
+    val thirdUpdates = spark.createDataFrame(thirdUpdateData).toDF(columns: _*)
+    thirdUpdates.write.format("hudi").
+      option(RECORDKEY_FIELD.key(), "key").
+      option(PRECOMBINE_FIELD.key(), "ts").
+      option(TABLE_TYPE.key(), tableType).
+      option(OPERATION.key(), "upsert").
+      options(opts).
+      mode(SaveMode.Append).
+      save(basePath)
+
+    val thirdDeletesData = Seq(
+      (10, "4", "rider-D4", "driver-D4", 34.15, 6),
+      (0, "1", "rider-X", "driver-X", 19.10, 8))
+    val thirdDeletes = spark.createDataFrame(thirdDeletesData).toDF(columns: _*)
+    thirdDeletes.write.format("hudi").
+      option(RECORDKEY_FIELD.key(), "key").
+      option(PRECOMBINE_FIELD.key(), "ts").
+      option(TABLE_TYPE.key(), tableType).
+      option(OPERATION.key(), "delete").
+      options(opts).
+      mode(SaveMode.Append).
+      save(basePath)
+
+    val fourUpdateData = Seq((-9, "4", "rider-DDDD", "driver-DDDD", 20.00, 1))
```

Comment:
lets do a round of validation before compaction. and once after compaction.

Reply:
and lets validate that in case of MOR table, compaction has not kicked in while we do the first validation. 

and post compaction, lets validate that 1 compaction commit is seen.

Reply:
Added a validation before compaction.